### PR TITLE
[Jerrold] Duke Increments

### PR DESCRIPTION
**Pull request by Wingedevil**

### CODE REVIEW

**Comment by yewon0303** [comment]
Your use of .useDelimiter() for scanner function is interesting and I've not seen many people use that for Scanner, it's unique!
I see that you have some class variables used to represent formats for the UI (PRINT_INDENT, PRINT_HORIZONTAL_LINE), that's nice! But I also see a few of the same `ui.write(String.format("%s%s\n", PRINT_INDENT, PRINT_HORIZONTAL_LINE);` being repeatedly used. Would you want to declare a new variable for that?

I also see that your src/main/java/Duke.java file is a little long. Have you considered having another class to deal with the commands input by the user?

The rest looks good! 👍

**Comment by kerwin97** [comment]
You Parser class is really clean and simple! Would love to learn from you! Could really tell you put in a lot of effort abstracting your functions! Your javaDoc comments are really informative too.
Looks good!

File: src/main/java/Duke.java [hunk 1]
```diff
@@ -1,10 +1,489 @@
...
+    private Scene scene;
+
     public static void main(String[] args) {
-        String logo = " ____        _        \n"
-                + "|  _ \\ _   _| | _____ \n"
-                + "| | | | | | | |/ / _ \\\n"
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        // new Duke(System.in, System.out, "data/duke.txt").run();
```
Comment (jon-chua) [inline src/main/java/Duke.java:49]: Commented/unused code can be removed. 😄

File: src/main/java/Duke.java [hunk 2]
```diff
@@ -1,10 +1,489 @@
...
+    }
+
+    // input expected: "delete n"
+    // deletes task n
+    // throws duke exception if there is no task n
+    private void delete(String[] args) throws IOException, DukeException {
+        if (args.length != 2){
+            throw new DukeException("\u2639 OOPS!!! Done function needs exactly one argument.");
+        }
+        if (!args[1].matches("\\d{1,9}")){
```
Comment (jon-chua) [inline src/main/java/Duke.java:208]: Try to use a comment to explain this regex expression? Also, this regex appears to match for numbers only, hence the thrown DukeException should be more descriptive (e.g. argument is not a number)

File: src/main/java/Duke.java [hunk 3]
```diff
@@ -1,10 +1,489 @@
...
+     * @throws DukeException
+     */
+    // input expected: "done n"
+    // marks task n as done
+    // throws duke exception if there is no task n
+    private void done(String[] args) throws IOException, DukeException {
+        if (args.length != 2){
+            throw new DukeException("\u2639 OOPS!!! Done function needs exactly one argument.");
+        }
+        if (!args[1].matches("\\d{1,9}")){
```
Comment (jon-chua) [inline src/main/java/Duke.java:239]: Same as with above

File: src/main/java/Duke.java [hunk 4]
```diff
@@ -1,10 +1,489 @@
...
+     * @throws IOException if stream fed to Duke is not valid
+     */
+    // input expected: "list"
+    // input not checked
+    // prints out list of tasks to out stream
+    private void listList(String[] args) throws IOException {
+        printList(todoList);
+    }
+
+    private void printList(TaskList<Task> tl) throws IOException {
```
Comment (jon-chua) [inline src/main/java/Duke.java:266]: Try to follow the SLAP principle - printList() and the related ui methods can be refactored into a separate Ui class 😄

File: src/main/java/Duke.java [hunk 5]
```diff
@@ -1,10 +1,489 @@
...
+        @Override
+        public void flush(){
+            createLabelAndAdd(print, false, true);
+            print = "";
+        }
+    }
+}
+
+
+class DukeException extends Exception {
```
Comment (jon-chua) [inline src/main/java/Duke.java:498]: DukeException can be refactored into its own file too! 😄

File: src/main/java/Storage.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        Object rtv = null;
+        try {
+            FileInputStream fileIn = new FileInputStream(file);
+            ObjectInputStream in = new ObjectInputStream(fileIn);
+            rtv = in.readObject();
+            in.close();
+            fileIn.close();
+        } catch (IOException | ClassNotFoundException e) {
+            // e.printStackTrace();
+        } finally {
```
Comment (jon-chua) [inline src/main/java/Storage.java:47]: Empty finally blocks can be removed! 😸

File: src/main/java/Ui.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+    /**
+     * Closes the streams associated with this UI.
+     */
+    public void close() {
+        try {
+            if (in != null) {
+                in.close();
+            }
+        } catch (IOException e1) {
```
Comment (jon-chua) [inline src/main/java/Ui.java:50]: Same as above

**Review by jon-chua — COMMENTED** [review]
Good job! 🥇 
Your code was well-documented and comprehensive.
However, I feel that you could have applied the SLAP principle more and refactor out relevant sections of code!
Overall, well done 👍

**Pull request closed without merging**